ledger_test.go

```diff
@@ -0,0 +1,204 @@
+
+	others := []*TestLedger{}
+	for i := 0; i < 5; i++ {
+		others = append(others, testnet.AddNode(t, 0))
+	}
+
+	testnet.WaitForConsensus(t)
+
+	// Alice attempt to pay Bob more than what
+	// she has in her wallet
+	assert.NoError(t, txError(alice.Pay(bob, 1000001)))
+
+	testnet.WaitForConsensus(t)
+
+	// Bob should not receive the tx amount
+	assert.EqualValues(t, 100, bob.Balance())
+
+	// Alice should have paid for gas even though the tx failed
+	aliceBalance := alice.Balance()
+	assert.True(t, aliceBalance > 0)
+	assert.True(t, aliceBalance < 1000000)
+
+	// Everyone else should see the updated balance of Alice and Bob
+	for _, node := range others {
+		assert.EqualValues(t, aliceBalance, node.BalanceOfAccount(alice))
+		assert.EqualValues(t, 100, node.BalanceOfAccount(bob))
+	}
+}
+
+func TestLedger_Gossip(t *testing.T) {
+	testnet := NewTestNetwork(t)
+	defer testnet.Cleanup()
+
+	alice := testnet.AddNode(t, 1000000)
+	bob := testnet.AddNode(t, 100)
+
+	for i := 0; i < 5; i++ {
+		testnet.AddNode(t, 0)
+	}
+
+	testnet.WaitForConsensus(t)
+
+	assert.NoError(t, txError(alice.Pay(bob, 1237)))
+
+	testnet.WaitForConsensus(t)
+
+	// When a new node joins the network, it will eventually receive
+	// all transactions in the network.
+	charlie := testnet.AddNode(t, 0)
+	timeout := time.NewTimer(15 * time.Second)
+	ticker := time.NewTicker(200 * time.Millisecond)
+
+	for {
+		select {
+		case <-timeout.C:
+			t.Fatal("test timed out")
+
+		case <-ticker.C:
+			if charlie.BalanceOfAccount(alice) == alice.Balance() &&
+				charlie.BalanceOfAccount(bob) == bob.Balance() {
+				return
+			}
+		}
+	}
+}
+
+func TestLedger_Stake(t *testing.T) {
+	testnet := NewTestNetwork(t)
+	defer testnet.Cleanup()
+
+	alice := testnet.AddNode(t, 1000000)
+
+	others := []*TestLedger{}
+	for i := 0; i < 5; i++ {
+		others = append(others, testnet.AddNode(t, 0))
+	}
+
+	testnet.WaitForConsensus(t)
+
+	assert.NoError(t, txError(alice.PlaceStake(9001)))
+	testnet.WaitForConsensus(t)
+
+	assert.EqualValues(t, 9001, alice.Stake())
+
+	// Alice balance should be balance-stakeAmount-gas
+	aliceBalance := alice.Balance()
+	assert.True(t, aliceBalance < 1000000-9001)
+
+	// Everyone else should see the updated balance of Alice
+	for _, node := range others {
+		assert.EqualValues(t, aliceBalance, node.BalanceOfAccount(alice))
+		assert.EqualValues(t, alice.Stake(), node.StakeOfAccount(alice))
+	}
+
+	assert.NoError(t, txError(alice.WithdrawStake(5000)))
+	testnet.WaitForConsensus(t)
+
+	assert.EqualValues(t, 4001, alice.Stake())
+
+	// Withdrawn stake should be added to balance
+	oldBalance := aliceBalance
+	aliceBalance = alice.Balance()
+	assert.True(t, aliceBalance > oldBalance)
+
+	// Everyone else should see the updated balance of Alice
+	for _, node := range others {
+		assert.EqualValues(t, aliceBalance, node.BalanceOfAccount(alice))
+		assert.EqualValues(t, alice.Stake(), node.StakeOfAccount(alice))
+	}
+}
+
+func TestLedger_Reward(t *testing.T) {
```

Comment:
This test takes more than 20 seconds (on my local machine) to run because it has to go through several consensus rounds to generate the rewards. Not sure if there's a better way to test the rewards generation.